engine/access/ingestion2/forest_manager.go

```diff
@@ -0,0 +1,369 @@
+package ingestion2
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/optimistic_sync"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/queue"
+	"github.com/onflow/flow-go/utils/logging"
+	"github.com/rs/zerolog"
+)
+
+const (
+	// forestManagerCheckInterval is the interval at which the forest manager checks for new
+	// processable results. This is used to ensure that the forest manager makes progress even
+	// if block finalization halts.
+	//
+	// This value should be greater than the block production rate.
+	forestManagerCheckInterval = 10 * time.Second
+)
+
+// BlockStatus represents the state of a block in the forest.
+type BlockStatus int
+
+const (
+	// BlockStatusSealed indicates the block has been sealed
+	BlockStatusSealed BlockStatus = iota
+	// BlockStatusFinalized indicates the block is finalized
+	BlockStatusFinalized
+	// BlockStatusCertified indicates the block is certified
+	BlockStatusCertified
+)
+
+// String returns the string representation of the block state
+func (bs BlockStatus) String() string {
+	switch bs {
+	case BlockStatusSealed:
+		return "sealed"
+	case BlockStatusFinalized:
+		return "finalized"
+	case BlockStatusCertified:
+		return "certified"
+	default:
+		return "unknown"
+	}
+}
```

Comment:
In my head, it would be natural to store this status with the execution results -- because in this context, we use it mostly as a descriptor for the result than the block.

Reply:
that's a good point, and I think this would simplify some of the tracking in the result forest as well.